### DISCUSSION TITLE
Is it possible to selectively use different custom notifications?

### DISCUSSION THREAD

**Discussion by citroid** [filamentphp/filament#5551] [Help] [answered]
Hi,
I would like to implement a custom notification that is displayed to the user before resource creation when possible duplicates have been found. 
This notification should contain a list with links to these duplicate resources, which is why I need to customize the notification.
I followed this [documentation](https://filamentphp.com/docs/2.x/notifications/customizing-notifications) and created both a notification view and a notification object.
Now when I try to use my custom notification by calling `DuplicateApplicationNotification::make()->send();` I still get the default notification instead.
According to the documentation I have to add the following line to the boot() method of a service provider:
`$this->app->bind(BaseNotification::class, `DuplicateApplicationNotification::class);`
This actually works, but it seems to replace the normal notification with my custom notification in general.
It does not allow me to use it selectively, which would be my preferred approach.
Hence, my question: Is it possible to use different notifications selectively and if so, what would I have to add in my code to achieve that?

**Comment by zepfietje** [comment, marked as answer]
You can achieve this by creating a custom notification class: https://filamentphp.com/docs/2.x/notifications/customizing-notifications#custom-notification-object.

In the class you will have to specify which view to use using a setter for example.

**Reply by citroid** [reply]
Thank you for the quick response!
I tried to follow your suggestion, but I always get the default notification shown instead of the custom one.
So far I did:

1. Add a function setView() to my CustomNotification class:
```
public function setView(String $notificationType): static
    { 
        // For "DuplicateApplicationNotification" change view
        if ($notificationType == 'DuplicateApplicationNotification') {
            Log::info("Set view for DuplicateApplicationNotification");
            $this->view = 'filament.resources.notifications.duplicate-application-notification';
            $this->viewIdentifier = 'duplicate-application-notification';
        }
        return $this;
    }
```
2. Add this line to the boot() function of my AppServiceProvider:
`$this->app->bind(BaseNotification::class, CustomNotification::class);`
3. Creating the CustomNotification this way:
```
CustomNotification::make()
                ->setView('DuplicateApplicationNotification')
```
The log output confirms that setView() actually enters the if branch, but the change of "view" seems to have no effect.
Am I missing something obvious here?

**Reply by zepfietje** [reply]
Going through the source code now, I think we might still need to add support for this.
Notification actions do already support dynamically setting the view.

You might want to have a look at the `Filament\Notifications\Actions\Action` class, specifically the `toArray`, `fromArray` and `isViewSafe` methods.
Feel free to create a PR already! 😄

**Reply by citroid** [reply]
I noticed that this has been addressed in a recent [PR](https://github.com/filamentphp/filament/pull/5739) (which I much appreciate by the way).
Unfortunately I'm struggling in using this correctly (I still get the default notification view) which is due to the "safeViews" concept I suspect.
My custom notification view is stored under "resources/views/filament/resources/application-resource/notifications/duplicate-application-notification.blade.php" and I try to set it as follows from the `CreateRecord` implementation of my Filament resource:

```
protected function beforeCreate(): void {
    // ...
    Notification::make()
        ->warning()
        ->safeViews('filament.resources.application-resource.notifications.duplicate-application-notification')
        ->view('filament.resources.application-resource.notifications.duplicate-application-notification')
        ->viewData([
            'test' => 'viewData Test'
         ])
         ->title('Possible duplicates found')
         ->body('Detailed description text goes here.')
         ->persistent()
         ->send();
    $this->halt();
}
```

Could you please give me a hint how the custom view has to be set so it passes the `isViewSafe` check?

**Reply by zepfietje** [reply]
The safe views property isn't stored on the notification when it's converted to an array, since that would break the security. Therefore, you either have to create a custom component or (easier option)  configure the safe views for all notifications in a service provider. The PR contained an example: https://github.com/filamentphp/filament/pull/5739#issuecomment-1438124517.

**Reply by citroid** [reply]
Thank you, that did the trick!
It works fine now when I pass an array with string values for `viewData`.
Creating an array of Actions and passing it in as
```
Notification::make()
    // ...
    ->viewData([
        'duplicate_actions' => $duplicateActions
     ])
     // ...
```
leads to an error ("Serialization of 'Closure' is not allowed" in \Illuminate\Session\Store on the save() method).
Is this intentional as I thought the `viewData` array would accept arbitrary data?
Please let me know if more information or code examples are required.

**Reply by zepfietje** [reply]
Are you passing a closure as data? Think that doesn't work.

**Reply by citroid** [reply]
Hmm since the exception says so apparently I do, but I'm not aware of where that comes from.
I create my Actions array this way:
```
$duplicateActions = [];
// Retrieve possible duplicates from Eloquent model like so:
$duplicates = Application::where(...)->get();
foreach ($duplicates as $duplicate) {
    $duplicateUrl = env('APP_URL') . '/admin/applications/' . $duplicate->id;
    $action = Action::make('view_' . $duplicate->id)
        ->url($duplicateUrl, shouldOpenInNewTab: true)
        ->label("Application filed " . $duplicate->created_at);
    array_push($duplicateActions, $action);
}
```
Then I add `$duplicateActions` to `viewData` as posted above.
And in my view I output these Actions like this:
```
@if ($viewData = $getViewData())
    @if (isset($viewData['duplicate_actions']))
        @foreach ($viewData['duplicate_actions'] as $duplicateAction)
                <x-notifications::actions.action
                    :action="$duplicateAction"
                    component="notifications::link"
                    :icon-position="$getIconPosition()"
                    class="filament-notifications-link-action"
                >
                    {{ $duplicateAction->getLabel() }}
                </x-notifications::actions.action>
        @endforeach
    @endif
@endif
```

**Reply by zepfietje** [reply]
Why are you passing actions to view data? Actions should be set using the existing `actions()` setter.

**Reply by citroid** [reply]
Actually that was my original approach.
But if I pass them to actions (together with actions for a "confirm" and "close" button which also need to be displayed) the actions will be placed on the same line (from left to right) and that doesn't work well with the available space within the notification view.
Therefore I would like to place the link actions to the duplicates below each other and only the final line should contain confirm and close button on the same line.
I was hoping to achieve that with a custom view, because I don't see how I could display the actions differently if I pass them all to `actions()`.

Now if I think about it again a good alternative could be to only have one link action which would open the list view (table) of the resource showing only the records of a given set of IDs (those of the duplicates).
If I define a custom filter like this example from the [docs](https://filamentphp.com/docs/2.x/tables/filters#custom-filter-forms) it might work. The only drawback then would be the filter also being displayed in the list view where it is not really useful.

**Reply by zepfietje** [reply]
Sounds like you should create a custom notification view. Check the original Filament notification view to see how to retrieve actions. You're totally free to style them in whatever way you like.
